torch/csrc/jit/codegen/cuda/transform_replay.cpp

```diff
@@ -687,134 +687,430 @@ std::deque<TensorView*> producersFor(TensorView* tv) {
+  TensorView* from;
+  TensorView* to;
 
-  auto pairwiseMap = PairwiseRootDomainMap(producer_tv, consumer_tv);
-  auto replayed_producer = TransformReplay::replayPasC(
-      producer_tv, consumer_tv, consumer_pos_it->second, pairwiseMap);
+  std::vector<RootIDInfo> root_id_info_from;
+  std::vector<RootIDInfo> root_id_info_to;
+};
 
-  auto producer_root = producer_tv->getMaybeRFactorDomain();
-  auto replayed_domain = replayed_producer.first->domain();
+// l < r means l contains a smaller amount of information about the starting
+// tensor than r.
+bool operator<(const NextHopInfo& l, const NextHopInfo& r) {
+  if (l.root_id_info_to.size() != r.root_id_info_to.size()) {
+    return l.root_id_info_to.size() < r.root_id_info_to.size();
+  }
+  size_t l_complete = std::count_if(
+      l.root_id_info_to.begin(),
+      l.root_id_info_to.end(),
+      [](const RootIDInfo& i) { return i.is_complete; });
+  size_t r_complete = std::count_if(
+      r.root_id_info_to.begin(),
+      r.root_id_info_to.end(),
+      [](const RootIDInfo& i) { return i.is_complete; });
+  return l_complete < r_complete;
+}
 
-  // Find the number of root IDs involved in the transformation
-  auto dep_vals = DependencyCheck::getAllValsBetween(
-      {producer_root.begin(), producer_root.end()},
-      {replayed_domain.begin(),
-       replayed_domain.begin() + replayed_producer.second});
+// l > r means l contains a bigger amount of information about the starting
+// tensor than r.
+bool operator>(const NextHopInfo& l, const NextHopInfo& r) {
+  return r < l;
+}
 
-  std::unordered_set<Val*> dep_vals_set{dep_vals.begin(), dep_vals.end()};
+// l == r means it is hard to tell which one of then contains more information
+bool operator==(const NextHopInfo& l, const NextHopInfo& r) {
+  return !(r < l) && !(l < r);
+}
 
-  auto n_transformed_root_dims = std::count_if(
-      producer_root.begin(),
-      producer_root.end(),
-      [&dep_vals_set](IterDomain* root_id) {
-        return dep_vals_set.find(root_id) != dep_vals_set.end();
-      });
+std::vector<RootIDInfo> getStartingRootIDInfo(TensorView* tv) {
+  std::vector<RootIDInfo> result;
+  auto root_domain = tv->getRootDomain();
+  result.reserve(root_domain.size());
+  for (auto id : root_domain) {
+    result.emplace_back(id);
+  }
+  return result;
+}
 
-  if (replayed_pos.find(producer_tv) != replayed_pos.end()) {
-    if (n_transformed_root_dims < n_replayed_root_dims.at(producer_tv) ||
-        (n_transformed_root_dims == n_replayed_root_dims.at(producer_tv) &&
-         replayed_producer.second <= replayed_pos.at(producer_tv))) {
-      return false; // NOLINT(clang-analyzer-cplusplus.NewDeleteLeaks)
+// Infer the compute-at position from the information of preserved reference
+// root ID.
+//
+// TODO:
+// I think I need to modify TransformReplay to add a new interface to specify
+// the root domains, instead of a position in the leaf domain. With the new
+// interface, this function will not be needed.
```

Comment:
This is due to a mismatch between `TransformReplay` and `TransformPropagator`. The former replays the first `N` leaf IDs, whereas the latter replays everything from the starting reference tensor. I suspect we would want the behavior of `TransformReplay`, but not sure.